mlrun/datastore/redis.py

```diff
@@ -0,0 +1,129 @@
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import os
+
+import redis
+import redis.cluster
+from storey.redis_driver import RedisType
+
+import mlrun
+
+from .base import DataStore
+
+
+class RedisStore(DataStore):
+    """
+    Partial implementation of DataStore over the Redis KV store
+
+    - no support for filesystem
+    - key and value sizes are limited to 512MB
+    """
+
+    def __init__(
+        self,
+        parent,
+        schema,
+        name,
+        endpoint="",
+    ):
+        super().__init__(parent, name, schema, endpoint)
+        self.headers = None
+
+        self.endpoint = self.endpoint or mlrun.mlconf.redis_url
+
+        if self.endpoint.startswith("rediss://"):
+            self.endpoint = self.endpoint[len("rediss://") :]
+            self.secure = True
+        elif self.endpoint.startswith("redis://"):
+            self.endpoint = self.endpoint[len("redis://") :]
+            self.secure = False
+        else:
+            raise NotImplementedError(f"invalid endpoint: {endpoint}")
+
+        self._redis_url = f"{schema}://{self.endpoint}"
+
+        self._redis = None
+
+        if os.environ.get("MLRUN_REDIS_TYPE") == "cluster":
+            self._redis_type = RedisType.CLUSTER
+        else:
+            self._redis_type = RedisType.STANDALONE
+
+    @property
+    def redis(self):
+        if self._redis is None:
+            if self._redis_type is RedisType.STANDALONE:
+                self._redis = redis.Redis.from_url(
+                    self._redis_url, decode_responses=True
+                )
+            else:
+                self._redis = redis.cluster.RedisCluster.from_url(
+                    self._redis_url, decode_response=True
+                )
+        return self._redis
+
+    def get_filesystem(self, silent):
+        return None  # no support for fsspec
+
+    def upload(self, key, src_path):
+        with open(src_path, "rb") as f:
+            while True:
+                data = f.read(1000 * 1000)
+                if not data:
+                    break
+                self.redis.append(key, data)
+
+    def get(self, key, size=-1, offset=0):
+        if offset or size:
+            end_offset = size if size == -1 else offset + size - 1
+            return self.redis.getrange(key, offset, end_offset)
+        else:
+            return self.redis.get(key)
+
+    def put(self, key, data, append=None):
+        if append is None:
+            self.redis.set(key, data)
+        else:
+            self.redis.append(key, data)
+
+    def stat(self, key):
+        raise NotImplementedError()
+
+    def listdir(self, key):
+        """
+        list all keys with prefix key
+        """
+        response = []
+        for key in self.redis.scan_iter(key + "*"):
```

Comment:
Don't you need `/*` here? If for example I do `listdir('name')` I'll get back `name/object1` etc., but also `name-with-more-chars`

Reply:
since redis keys have no hierarchy and the datastore is generic, i didn't want to assume that there is directory hierarchy in place (and the datastore client can always provide the '/' postfix if it so chooses)
however, if you think things will  break for users of the datastore that assume that a dir structure always exist, then maybe indeed we should change this.

Reply:
It is pretty much standard to use `/` to emulate a directory structure. This is done in S3 for example - there is no real hierarchy of objects there, but the standard is to emulate it like this.
You can simply do what the `S3Store` does -
```
if not key.endswith("/"):
   key += "/"
```
Still, I suggest we stick to this convention if providing a `listdir` function.

Reply:
fixed as suggested